### ISSUE TITLE
Replace `if let tuple = tuple` by a let chain

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-clippy#17169]
### What it does

Replaces `if let (Some(x), Ok(y)) = (x, y)` by `if let Some(x) = x && Ok(y) = y`, i.e. when a tuple is constructed and then immediately matched against a destructuring pattern. Similarly for arrays.

Things like `if let (Some(x), Some(y) = (y, x)`, albeit questionable, should not trigger the lint.

### Advantage

- Separates independent checks
- Can short-circuit

### Drawbacks

- Can short circuit
- With default formatting, can increase the number of lines

### Example

```rust
if let (Some(x), Ok(y)) = (x, y) && let [Some(_), None] = [a, b] {
   ...
}
```

Could be written as:

```rust
if let Some(x) = x && let Ok(y) = y && let Some(_) = a && let None = b {
   ...
}
```

### Comparison with existing lints

It seems likely that `redundant_pattern_match` can trigger as a follow-up. 

Not a lint, but rust-analyzer's `unwrap_tuple` assist is the analog for assignments. Note that for assignments, `(x, y) = (y, x);` is entirely reasonable and idiomatic,

### Additional Context

This pattern may have been used as a partial workaround before let chains were available.

**Comment by arshalaromal** [comment]
Taking a look at this. Assuming we just check for side-effects first to avoid changing the short-circuit behavior?

**Comment by samueltardieu** [comment]
Can you exhibit examples where the side-effect free code would generate different code in release mode? For example, those two functions generate the exact same code:

```rust
#[inline(never)]
pub fn f(x: Option<u32>, y: Result<u32, &str>) -> Result<u32, ()> {
    if let (Some(x), Ok(y)) = (x, y) {
        Ok(x + y)
    } else {
        Err(())
    }
}

#[inline(never)]
pub fn g(x: Option<u32>, y: Result<u32, &str>) -> Result<u32, ()> {
    if let Some(x) = x && let Ok(y) = y {
        Ok(x + y)
    } else {
        Err(())
    }
}
```

**Comment by ronnodas** [comment]
@samueltardieu I'm not sure if this is what you were looking for but here is an example where the let-chain version executes fewer side-effects:
```rust
fn with_side_effect(text: &str) -> Option<u32> {
    println!("{}", text);
    None
}

pub fn main() {
    let none: Option<u32> = None;
    if let (Some(_x), Some(_y)) = (none, with_side_effect("hello")) {
    }
    if let Some(_x) = none && let Some(_y) = with_side_effect("hello") {
    }
}
```

**Comment by samueltardieu** [comment]
Of course when there are side effects we execute less code, but then this is no longer equivalent. What I'm interested in is, when there are *no* side effect, how the proposed replacement will make the code better.

**Comment by ronnodas** [comment]
Just replace the `with_side_effect()` function with something pure but computationally expensive.

Or perhaps you're saying that compiler optimizations take care of that already? Then consider an expensive computation that is cached, so from the compiler's perspective, does have side-effects.

Besides performance, I think the second version is clearer to read than the first, especially if `x` and `y` are obtained independently, say `x` is looked up in a hash map and `y` is the result of a separate computation. This is what I meant by "Separates independent checks" in the original post.

**Comment by samueltardieu** [comment]
> Besides performance, I think the second version is clearer to read than the first

Other people might think the opposite, and prefer to group related assignments together to mark them as such. Such a lint would probably be acceptable as a `restriction` lint which would need to be activated, but hardly as a warn-by-default or even a `pedantic` one as it is really a matter of taste.

**Comment by ronnodas** [comment]
I see what you're saying. What about cases where a further simplification lint would apply, like `redundant_pattern_match`? Maybe something along these lines makes more sense as an enhancement of that lint? It must be less contentious that 
```rust
if x.is_some() && y { ... }
```
is better than
```rust
if let (Some(_), true) = (x, y) { ... }
```
